src/features/side_by_side_wrap.rs

```diff
@@ -0,0 +1,1115 @@
+            self.line_segments.push(text);
+            self.len = len;
+        }
+        fn has_text(&self) -> bool {
+            self.len > LINEPREFIX.len()
+        }
+        fn text_len(&self) -> usize {
+            if self.len > LINEPREFIX.len() {
+                self.len - LINEPREFIX.len()
+            } else {
+                debug_assert!(false, "push or reset first");
+                0
+            }
+        }
+    }
+
+    let mut curr_line: CurrLine<S> = CurrLine {
+        line_segments: Vec::new(),
+        len: 0,
+    };
+
+    // Determine the background (diff) and color (syntax) of an inserted symbol.
+    let symbol_style = match inline_hint_style {
+        Some(style) => *style,
+        None => *fill_style,
+    };
+
+    let mut stack = line.into_iter().rev().collect::<Vec<_>>();
+
+    let line_limit_reached = |result: &Vec<_>| {
+        // If only the wrap symbol and no extra text fits, then wrapping is not possible.
+        let max_lines = if line_width <= INLINE_SYMBOL_WIDTH_1 {
+            1
+        } else {
+            wrap_config.max_lines
+        };
+
+        max_lines > 0 && result.len() + 1 >= max_lines
+    };
+
+    while !stack.is_empty() && !line_limit_reached(&result) && max_len > LINEPREFIX.len() {
+        let (style, text, graphemes) = stack
+            .pop()
+            .map(|(style, text)| (style, text, text.grapheme_indices(true).collect::<Vec<_>>()))
+            .unwrap();
+
+        let new_len = curr_line.len + graphemes.len();
+
+        let must_split = if new_len < max_len {
+            curr_line.push_and_set_len((style, text), new_len);
+            false
+        } else if new_len == max_len {
+            match stack.last() {
+                // Perfect fit, no need to make space for a `wrap_symbol`.
+                None => {
+                    curr_line.push_and_set_len((style, text), new_len);
+                    false
+                }
+                #[allow(clippy::identity_op)]
+                // A single '\n' left on the stack can be pushed onto the current line.
+                Some((next_style, nl)) if stack.len() == 1 && *nl == "\n" => {
+                    curr_line.push_and_set_len((style, text), new_len);
+                    // Do not count the '\n': + 0
+                    curr_line.push_and_set_len((*next_style, *nl), new_len + 0);
+                    stack.pop();
+                    false
+                }
+                _ => true,
+            }
+        } else if new_len == max_len + 1 && stack.is_empty() {
+            // If the one overhanging char is '\n' then keep it on the current line.
+            if !text.is_empty() && *text.as_bytes().last().unwrap() == b'\n' {
+                // Do not count the included '\n': - 1
+                curr_line.push_and_set_len((style, text), new_len - 1);
+                false
+            } else {
+                true
+            }
+        } else {
+            true
+        };
```

Comment:
`wrap_line` is a long, complex function. I think that, personally, I would factor out some helper functions such as `get_must_split` here (we spend 30 lines computing this boolean), and I think `split()` and `align()` methods below.

In conjunction with my comment above about the struct definition inside the function body, perhaps what I'm wondering is -- could it make sense to have a top level `LineWrapper` struct, holding the mutable data used in the wrapping algorithm, with `wrap_line` an associated function and additional helper functions `get_must_split()`, `split()`, and `align()`. I'm thinking that the actual wrapping algorithm would be much more apparent with a refactoring like that.